include/PLS/pls.h

```diff
@@ -32,20 +32,36 @@ typedef Eigen::Matrix<size_t, Eigen::Dynamic, 1> Colsz;
 typedef Eigen::RowVectorXi Rowi;
 typedef Eigen::Matrix<size_t, 1, Eigen::Dynamic> Rowsz;
 
+/**
+@brief Namespace for all Partial Least Square (PLS) related functions and classes.
+*/
 namespace PLS {
 
-    typedef std::vector<Mat2D> Residual;
+    // error data structure
+    typedef std::vector<Mat2D> ResidualData;
+    class Model; // forward declaration for Residual friendship
+
+    class Residual {
+        const ResidualData _residual;
```

Comment:
Why use a typedef for this?  This is probably at least somewhat philosophical, but vector<Mat2D> seems perfectly clear (and not much more verbose than ResidualData), whereas the typedef (1) makes it seem somehow more exotic than it is and (2) calling it "data" feels weird, since to a lot of people that means empirical observations, and even for our (I think we share this view) broader take that simulator output is still data, it feels a bit odd.  If you *really* want to use a typedef, maybe ResidualValues.

Reply:
I think of this a bit of future-proofing; I'm not convinced this is the best *actual* type to use - making it a typedef telegraphs (but doesn't enforce) that consumers shouldn't be picking at the internals, and keeps the PLS interface consistent(ly written, if not actually-typed) in the event of future changes to this structure.

Reply:
went with this approach